### PR TITLE
serve restic: Fix error handling

### PR DESCRIPTION
**Pull request by MichaelEischer**
#### What is the purpose of this change?
We've discovered in https://github.com/restic/restic/issues/4612 that rclone returns a "not found" error instead of an "internal server error" if listing a directory fails. This is very problematic as since restic 0.16.0 a "not found" error while listing a directory is treated like an empty directory. Consequently, the `prune` command can end up believing that there are no snapshots if an error occurs while listing the snapshots and wipe the entire repository.

The second commit modifies the error handling while retrieving a file, such that it only returns "not found" if there's a corresponding error. A future restic version will stop retrying "not found" errors (see https://github.com/restic/restic/pull/4520) and therefore needs to be able to distinguish permanent errors like "not found" from other probably temporary ones.

I haven't added any tests as I'm not sure how to best test the error handling here.

#### Was the change discussed in an issue or in the forum before?
No.

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [ ] I have added tests for all changes in this PR if appropriate.
- ~~[ ] I have added documentation for the changes if appropriate.~~
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

File: cmd/serve/restic/restic.go [hunk 1]
```diff
@@ -341,7 +341,11 @@ func (s *server) serveObject(w http.ResponseWriter, r *http.Request) {
 	o, err := s.newObject(r.Context(), remote)
 	if err != nil {
 		fs.Debugf(remote, "%s request error: %v", r.Method, err)
-		http.Error(w, http.StatusText(http.StatusNotFound), http.StatusNotFound)
+		if err == fs.ErrorObjectNotFound {
```
Comment (ncw) [inline cmd/serve/restic/restic.go:344]: This should probably use `errors.Is` as we do elsewhere in this code.

**Review by ncw — COMMENTED** [review]
This looks great - see comment inline.

Does the restic `TestBackendRESTExternalServer` tests test this case? We run these tests against rclone on the integration test server.

I think you could add a test here like this

https://github.com/rclone/rclone/blob/1045f5412804c73fb729e7d1ab0a99713aef75f9/cmd/serve/restic/restic_privaterepos_test.go#L68-L76

Not sure why the restic tests are in lots of different files but if you added a similar test to that file that would be perfect.

File: cmd/serve/restic/restic.go (hunk 1 above)
Comment (MichaelEischer) [reply cmd/serve/restic/restic.go]: I've also added a commit that replaces the err comparison used when deleting a file.

**Comment by MichaelEischer** [comment]
I've added two small tests to check the new error handling behavior. PTAL

**Review by ncw — APPROVED** [review]
Thank you, that looks great. I'll merge it now :-)

**Pull request merged**